sql/log.cc

```diff
@@ -2908,6 +2901,33 @@ Please consider archiving some logs.", next, (MAX_LOG_UNIQUE_FN_EXT - next));
   DBUG_RETURN(error);
 }
 
+/** Fail if binlog_error_action is ABORT_SERVER ~~or above~~,
+    or complain the previous `fatal_log_error` message otherwise */
+static void fatal_log_error(const char *name, int error_id= errno)
+{
+  if (binlog_error_action >= ABORT_SERVER)
+  {
+    THD *thd= current_thd;
+    /*
+      On fatal error when code enters here we should forcefully clear the
+      previous errors so that a new critical error message can be pushed
+      to the client side.
+    */
+    thd->clear_error();
+    static const char *abort_error=
+      "Either disk is full or file system is read only while opening the binlog. "
+      "Aborting the server";
+    //XXX: MySQL needs both error calls. mysql/mysql-server@3b6b4bf8c5
+    my_error(ER_BINLOG_LOGGING_IMPOSSIBLE, MYF(0), abort_error);
+    sql_print_error("Could not use %s for logging (error %d). %s", name, error_id, abort_error);
```

Comment:
mysql/mysql-server@3b6b4bf8c5 amended the second error call.

I wonder, what’s the difference between `my_error` and `sql_print_error`, and why did they (if not *us*) need to error twice?

I know that `my_error` receives its format template from `errmsg-utf8.txt` while `sql_print_error` uses a given template. (I know this from `#3360`.) This is why they have different error messages here.

https://github.com/MariaDB/server/blob/ebbbe9d96023ed4f12409abf9c363d4ea9113089/sql/share/errmsg-utf8.txt#L8311-L8317
(Yes, `ER_BINLOG_LOGGING_IMPOSSIBLE` already exists.)

Reply:
Good question, I'm actually not sure.. I imagine we should only need `my_error` , though we should extend `abort_error` with the 

> Could not use %s ..

details.